### PR TITLE
FAT-2479: Implement e-2-e automation of test case C356772

### PR DESCRIPTION
**Pull request by re-jas**
https://foliotest.testrail.io/index.php?/cases/view/356772&group_by=cases:priority_id&group_order=asc&display_deleted_cases=0&group_id=3

### CODE REVIEW

File: cypress/integration/checkout/checkout.basic.spec.js [hunk 1]
```diff
@@ -1,54 +1,122 @@
...
-  let source;
+describe('Check Out - Actions ', () => {
+  const userData = {
+    permissions: [
+      permissions.checkinAll.gui,
+      permissions.uiCirculationViewCreateEditDelete.gui,
+      permissions.uiUserCreate.gui,
+      permissions.inventoryAll.gui,
+      permissions.checkoutCirculatingItems.gui,
+    ],
```
Comment (penfriend) [inline cypress/integration/checkout/checkout.basic.spec.js:24]: Please move permissions to the place where you use it one time to the parameter of the method  cy.createTempUser

File: cypress/integration/checkout/checkout.basic.spec.js [hunk 2]
```diff
@@ -1,54 +1,122 @@
...
+
+        cy.login(userData.username, userData.password);
+      });
+  });
 
-    cy.checkOutItem(Cypress.env('users')[0].barcode, ITEM_BARCODE);
+  after('Delete New Service point, Item and User', () => {
+    cy.visit('/checkin');
+    CheckInActions.checkInItemGui(itemData.barcode);
+    CheckInPane.checkResultsInTheRow(checkInResultsData);
```
Comment (penfriend) [inline cypress/integration/checkout/checkout.basic.spec.js:100]: The purpose of this autotest doesn't contains the testing the check in results. Please remove this check  CheckInPane.checkResultsInTheRow(checkInResultsData);

File: cypress/integration/checkout/checkout.basic.spec.js [hunk 3]
```diff
@@ -1,54 +1,122 @@
...
+        UserEdit.addServicePointViaApi(servicePoint.id, userData.userId, servicePoint.id);
+
+        cy.login(userData.username, userData.password);
+      });
+  });
 
-    cy.checkOutItem(Cypress.env('users')[0].barcode, ITEM_BARCODE);
+  after('Delete New Service point, Item and User', () => {
+    cy.visit('/checkin');
+    CheckInActions.checkInItemGui(itemData.barcode);
```
Comment (penfriend) [inline cypress/integration/checkout/checkout.basic.spec.js:99]: Usually in the before and after blocks we are using api methods. So here I recommend you to use the check in items via api. In this case you don't need this two lines of code. Please rewrite this two line into one line using the api method to check in the item

File: cypress/integration/checkout/checkout.basic.spec.js [hunk 4]
```diff
@@ -1,54 +1,122 @@
...
+        userData.username = userProperties.username;
+        userData.password = userProperties.password;
+        userData.userId = userProperties.userId;
+        userData.barcode = userProperties.barcode;
+        userData.firstName = userProperties.firstName;
+      })
+      .then(() => {
+        UserEdit.addServicePointViaApi(servicePoint.id, userData.userId, servicePoint.id);
+
+        cy.login(userData.username, userData.password);
```
Comment (penfriend) [inline cypress/integration/checkout/checkout.basic.spec.js:93]: Please add two parameters to this function including path and waiter. the path should be the path to checkout app. You can import it using the import topMenu from '../../support/fragments/topMenu'; and please find the waiter in the related fradment file

File: cypress/integration/checkout/checkout.basic.spec.js [hunk 5]
```diff
@@ -1,54 +1,122 @@
...
+      defaultLocation.libraryId,
+      defaultLocation.id
+    );
+    ServicePoints.deleteViaApi(servicePoint.id);
+  });
 
-    cy.verifyItemCheckOut();
+  it('C356772 An active user with barcode can Check out item (vega)', { tags: [TestTypes.smoke, devTeams.vega] }, () => {
+    cy.visit('/checkout');
+    CheckOutActions.checkIsInterfacesOpened();
```
Comment (penfriend) [inline cypress/integration/checkout/checkout.basic.spec.js:115]: if you add the waiter in the login function this line will be redundant

File: cypress/integration/checkout/checkout.basic.spec.js [hunk 6]
```diff
@@ -1,54 +1,122 @@
...
+      defaultLocation.campusId,
+      defaultLocation.libraryId,
+      defaultLocation.id
+    );
+    ServicePoints.deleteViaApi(servicePoint.id);
+  });
 
-    cy.verifyItemCheckOut();
+  it('C356772 An active user with barcode can Check out item (vega)', { tags: [TestTypes.smoke, devTeams.vega] }, () => {
+    cy.visit('/checkout');
```
Comment (penfriend) [inline cypress/integration/checkout/checkout.basic.spec.js:114]: if you add the waiter in the login function this line will be redundant

File: cypress/integration/checkout/checkout.basic.spec.js [hunk 7]
```diff
@@ -1,54 +1,122 @@
...
+    CheckOutActions.checkIsInterfacesOpened();
+    CheckOutActions.checkOutUser(userData.barcode);
+    CheckOutActions.checkUserInfo(userData, 'staff');
+    CheckOutActions.checkOutItem(itemData.barcode);
+    // cy.verifyItemCheckOut();
```
Comment (penfriend) [inline cypress/integration/checkout/checkout.basic.spec.js:119]: we usually don't merge pull requests with commented line's of code

File: cypress/integration/checkout/checkout.basic.spec.js [hunk 8]
```diff
@@ -1,54 +1,122 @@
...
+    );
+    ServicePoints.deleteViaApi(servicePoint.id);
+  });
 
-    cy.verifyItemCheckOut();
+  it('C356772 An active user with barcode can Check out item (vega)', { tags: [TestTypes.smoke, devTeams.vega] }, () => {
+    cy.visit('/checkout');
+    CheckOutActions.checkIsInterfacesOpened();
+    CheckOutActions.checkOutUser(userData.barcode);
+    CheckOutActions.checkUserInfo(userData, 'staff');
```
Comment (penfriend) [inline cypress/integration/checkout/checkout.basic.spec.js:117]: Are you sure that you always have the patron group for the user with the name 'staff'? To be sure you need add this parameter during the creation of the temp user.

File: cypress/support/fragments/check-out-actions/check-out-actions.js [hunk 9]
```diff
@@ -26,11 +26,11 @@ export default {
     ]);
   },
 
-  async checkUserInfo({ barcode, personal }, patronGroup = '0') {
+  async checkUserInfo({ barcode, username }, patronGroup = '0') {
```
Comment (penfriend) [inline cypress/support/fragments/check-out-actions/check-out-actions.js:29]: This method can be used in other autotests. So changing this method can affect them. Please don't change this method in this way, you can add your own new method to the test this case.
Also I have doubts if the username is shown always. Need to recheck the cases.

File: cypress/integration/checkout/checkout.basic.spec.js [hunk 10]
```diff
@@ -1,54 +1,122 @@
...
+    barcode: generateItemBarcode(),
+    instanceTitle: `Instance ${getRandomPostfix()}`,
+  };
+  let defaultLocation;
+  const servicePoint = ServicePoints.getDefaultServicePointWithPickUpLocation('autotest basic checkin', uuid());
+  const checkInResultsData = ['Available', itemData.barcode];
 
-  beforeEach(() => {
-    cy.login(Cypress.env('diku_login'), Cypress.env('diku_password'));
+  before('Create New Item and New User', () => {
```
Comment (penfriend) [inline cypress/integration/checkout/checkout.basic.spec.js:34]: here you also create service point, location, etc.

File: cypress/integration/checkout/checkout.basic.spec.js [hunk 11]
```diff
@@ -1,54 +1,122 @@
...
+      permissions.checkoutCirculatingItems.gui,
+    ],
+  };
+  const itemData = {
+    barcode: generateItemBarcode(),
+    instanceTitle: `Instance ${getRandomPostfix()}`,
+  };
+  let defaultLocation;
+  const servicePoint = ServicePoints.getDefaultServicePointWithPickUpLocation('autotest basic checkin', uuid());
+  const checkInResultsData = ['Available', itemData.barcode];
```
Comment (penfriend) [inline cypress/integration/checkout/checkout.basic.spec.js:32]: you don't need this variable. Please remove it

**Review by penfriend — CHANGES_REQUESTED** [review]

**Comment by penfriend** [comment]
> [CLA assistant check](https://cla-assistant.io/folio-org/stripes-testing?pullRequest=579) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/folio-org/stripes-testing?pullRequest=579) before we can accept your contribution.
> 
> **Jasurbek Erkinov** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).
> You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/folio-org/stripes-testing?pullRequest=579) it.

Please check your email address

File: cypress/support/fragments/check-out-actions/check-out-actions.js [hunk 12]
```diff
@@ -81,5 +83,10 @@ export default {
 
   checkItem:(barcode) => {
     cy.expect(MultiColumnList({ id:'list-items-checked-out' }).find(HTML(including(barcode))).absent());
-  }
+  },
+
+  waitLoading:() => {
+    cy.expect(patronBarcodeField.exists());
+    cy.expect(itemBarcodeField.exists());
+  },
```
Comment (penfriend) [inline cypress/support/fragments/check-out-actions/check-out-actions.js:91]: we have this method waitLoading() created in the checkout fragment. Please use the method from the checkout fragment

File: cypress/support/fragments/check-out-actions/check-out-actions.js [hunk 13]
```diff
@@ -16,6 +16,8 @@ import {
 const modal = Modal('Confirm multipiece check out');
 const endSessionButton = Button('End session');
 const userPane = PaneContent({ id: 'patron-details-content' });
+const itemBarcodeField = TextField({ name:'item.barcode' });
+const patronBarcodeField = TextField({ name: 'patron.identifier' });
```
Comment (penfriend) [inline cypress/support/fragments/check-out-actions/check-out-actions.js:20]: so here these vars are redundant and the method waitLoading are redundant

File: cypress/integration/checkout/checkout.basic.spec.js [hunk 14]
```diff
@@ -1,54 +1,122 @@
...
 
-  it('Basic flow', function () {
-    cy.visit('/checkout');
+    cy.createTempUser(
+      [
+        permissions.checkinAll.gui,
+        permissions.uiCirculationViewCreateEditDelete.gui,
+        permissions.uiUserCreate.gui,
+        permissions.inventoryAll.gui,
+        permissions.checkoutCirculatingItems.gui,
```
Comment (penfriend) [inline cypress/integration/checkout/checkout.basic.spec.js:80]: I don't see here the permission Settings (Circ): Can create, edit and remove other settings

Comment (penfriend) [reply cypress/integration/checkout/checkout.basic.spec.js:80]: We have such precondition. Please add it to your test
Setting: Circulation > Other settings, Username checkbox = true

Comment (penfriend) [reply cypress/integration/checkout/checkout.basic.spec.js:80]: I don't see here the permission *Users: Can create new user;

File: cypress/integration/checkout/checkout.basic.spec.js [hunk 15]
```diff
@@ -1,54 +1,122 @@
...
+      defaultLocation.campusId,
+      defaultLocation.libraryId,
+      defaultLocation.id
+    );
+    ServicePoints.deleteViaApi(servicePoint.id);
+  });
 
-    cy.verifyItemCheckOut();
+  it('C356772 An active user with barcode can Check out item (vega)', { tags: [TestTypes.smoke, devTeams.vega] }, () => {
+    CheckOutActions.checkOutUser(userData.barcode);
```
Comment (penfriend) [inline cypress/integration/checkout/checkout.basic.spec.js:117]: here you use the barcode but you need to find a user using the username

File: cypress/integration/checkout/checkout.basic.spec.js [hunk 16]
```diff
@@ -1,54 +1,122 @@
...
+
+  after('Delete New Service point, Item and User', () => {
+    CheckInActions.checkinItemViaApi({
+      itemBarcode: itemData.barcode,
+      servicePointId: servicePoint.id,
+      checkInDate: new Date().toISOString(),
+    });
+    InventoryInstances.deleteInstanceAndHoldingRecordAndAllItemsViaApi(itemData.barcode);
+    UserEdit.changeServicePointPreferenceViaApi(userData.userId, [servicePoint.id]);
+    Users.deleteViaApi(userData.userId);
```
Comment (penfriend) [inline cypress/integration/checkout/checkout.basic.spec.js:106]: Actually we need to create two users according to the preconditions. One user should create another user (Created test active user with a username and barcode exist) and checkout the item to that user.

**Review by penfriend — CHANGES_REQUESTED** [review]

File: cypress/support/fragments/users/users.js [hunk 17]
```diff
@@ -38,4 +39,23 @@ export default {
...
+      Dropdown('Actions').find(Button()).click(),
+      Button({ id: 'clickable-newuser' }).click(),
+      TextField({ id: 'adduser_lastname' }).fillIn(userData.personal.lastName),
+      Select({ id: 'adduser_group' }).choose(userData.patronGroup),
+      TextField({ name: 'barcode' }).fillIn(userData.barcode),
+      TextField({ name: 'username' }).fillIn(userData.username),
+      TextField({ id: 'adduser_email' }).fillIn(userData.personal.email),
+      Button({ id: 'clickable-save' }).click()]).then(() => {
+      cy.intercept('/users').as('user');
+      return cy.wait('@user', { timeout: 30000 }).then((xhr) => xhr.response.body.id);
```
Comment (penfriend) [inline cypress/support/fragments/users/users.js:54]: please decrease the timeout to 3000-4000

File: cypress/integration/checkout/checkout.basic.spec.js [hunk 18]
```diff
@@ -1,54 +1,139 @@
...
 
-describe('Check Out', () => {
-  const ITEM_BARCODE = generateItemBarcode();
-  let source;
+describe('Check Out - Actions ', () => {
+  const userData = {
+    group: 'staff',
+    personal: {},
+  };
+  const testActiveUser = DefaultUser.defaultUiPatron.body;
```
Comment (penfriend) [inline cypress/integration/checkout/checkout.basic.spec.js:24]: please copy the object this way
{... DefaultUser.defaultUiPatron.body}

File: cypress/support/fragments/check-out-actions/check-out-actions.js [hunk 19]
```diff
@@ -18,9 +18,9 @@ const endSessionButton = Button('End session');
 const userPane = PaneContent({ id: 'patron-details-content' });
 export default {
   modal,
-  checkOutUser(userBarcode) {
+  checkOutUser(userBarcode, otherParametr) {
```
Comment (penfriend) [inline cypress/support/fragments/check-out-actions/check-out-actions.js:21]: otherParameter

File: cypress/support/fragments/check-out-actions/check-out-actions.js [hunk 20]
```diff
@@ -18,9 +18,9 @@ const endSessionButton = Button('End session');
...
     return cy.do([
-      TextField('Patron identifier').fillIn(userBarcode),
+      TextField('Patron identifier').fillIn(otherParametr || userBarcode),
```
Comment (penfriend) [inline cypress/support/fragments/check-out-actions/check-out-actions.js:23]: otherParameter

File: cypress/support/fragments/users/users.js (hunk 17 above)
Comment (re-jas) [reply cypress/support/fragments/users/users.js:54]: I tried to set the less waiter, but in this case there is not enough time and it gives an error, so I had to set the more value. The problem is that creating a user through UI. It takes a decent amount of time to process requests

**Review by penfriend — APPROVED** [review]

**Review by MaksimTerekhov — APPROVED** [review]

**Pull request merged**